### ISSUE TITLE
Usage of `hash` causes multiple indices to be created when PYTHONHASHSEED changes

### ISSUE THREAD

**Issue by akx** [pudo/dataset#130]
Since index name creation in [`dataset/persistence/table.py`](https://github.com/pudo/dataset/blob/master/dataset/persistence/table.py#L275) uses the `hash` function instead of a crypto hash function, hash seed randomization will cause multiple equivalent indexes to be created.

Hash seed randomization is enabled

* on Python 2 with the `-R` interpreter argument
* on Python 2 with the `PYTHONHASHSEED=random` environment variable
* [on Python 3.3+ by default](https://docs.python.org/3/using/cmdline.html#cmdoption-R)

(I personally bumped into this using Dataset with Python 3.4.)

**Comment by pudo** [comment]
Legit bug. Do you have a proposed work-around/fix?

**Comment by akx** [comment]
The obvious solution is to use a (truncated) cryptographic hash like

```python
sig = hashlib.md5('||'.join(columns)).hexdigest()[:16]
```

but that will naturally cause one round of duplicate indexes to be created for old Datasets.

An option to avoid this would be to have several choices of automatic names like

```python
if not name:
    col_names = '||'.join(columns)
    names = [
        'ix_%s_%s' % (self.table.name, sig)
        for sig in (
            abs(hash(col_names)),
            hashlib.md5(col_names).hexdigest()[:16]
        )
    ]
else:
    names = [name]

for name in names:
    if name in self.indexes:
        return self.indexes[name]

# Fall through and continue with the last `name`,
# i.e. the caller-entered or MD5-derived one
```

but that's starting to smell like it's getting a little complex for this?

**Comment by pudo** [comment]
As a non-breaking fix, I guess we'd check for the old ``hash()``-based index, wish for luck with the Gods of Random, and then use a hash when we need to create a new thing?

btw, why did you choose md5 in your example over sha1? Is it faster? I understand we don't care about it's 'safety' here.

**Comment by akx** [comment]
> As a non-breaking fix, I guess we'd check for the old hash()-based index, wish for luck with the Gods of Random, and then use a hashlib hash when we need to create a new thing?

That's what I was going for with the more complex code there, yeah. :)

No particular reason to go for MD5 over SHA1, both are just fine really.

**Comment by pudo** [comment]
Fixed in 61b33d9d4f3c1e05bb8e11efac2c9960a2e4a0d0 - time for 0.6.1, I think.

**Comment by akx** [comment]
Cheers! :)
